pkg/scheduler/queue/tree_queue_test.go

```diff
@@ -0,0 +1,480 @@
+	assert.True(t, root.IsEmpty())
+}
+
+// TestDequeueByPathUnbalancedTree checks dequeuing behavior from an unbalanced tree by path.
+//
+// Assertions are done one by one to illustrate and check the behaviors of dequeuing from
+// an unbalanced tree, where the same node will be dequeued from twice if the node remains
+// nonempty while its sibling nodes have been exhausted and deleted from the tree.
+func TestDequeueByPathUnbalancedTree(t *testing.T) {
+	root := makeUnbalancedTreeQueue(t)
+
+	// dequeue from root:2 until exhausted
+	path := QueuePath{"2"}
+	v := root.DequeueByPath(path)
+	assert.Equal(t, "root:2:0:val0", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:2:1:val0", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:2:0:val1", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:2:1:val1", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:2:1:val2", v)
+
+	// root:2 is exhausted;
+	v = root.DequeueByPath(path)
+	assert.Nil(t, v)
+	// root:2 and its two children root:2:0 and root:2:1 were deleted
+	assert.Nil(t, root.getNode(path))
+	assert.Equal(t, 2, len(root.childQueueMap))
+	assert.Equal(t, 2, len(root.childQueueOrder))
+	assert.Equal(t, 4, root.NodeCount())
+	// 5 of 10 items were dequeued
+	assert.Equal(t, 5, root.ItemCount())
+
+	// dequeue from root:1 until exhausted
+	path = QueuePath{"1"}
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:1:val0", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:1:0:val0", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:1:val1", v)
+
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:1:0:val1", v)
+
+	// root:1 is exhausted;
+	v = root.DequeueByPath(path)
+	assert.Nil(t, v)
+	// root:1 and its child root:1:0 were deleted
+	assert.Nil(t, root.getNode(path))
+	assert.Equal(t, 1, len(root.childQueueMap))
+	assert.Equal(t, 1, len(root.childQueueOrder))
+	assert.Equal(t, 2, root.NodeCount())
+	// 9 of 10 items have been dequeued
+	assert.Equal(t, 1, root.ItemCount())
+
+	// dequeue from root:0 until exhausted
+	path = QueuePath{"0"}
+	v = root.DequeueByPath(path)
+	assert.Equal(t, "root:0:val0", v)
+
+	// root:0 is exhausted;
+	v = root.DequeueByPath(path)
+	assert.Nil(t, v)
+	// root:0 was deleted
+	assert.Nil(t, root.getNode(path))
+	assert.Equal(t, 0, len(root.childQueueMap))
+	assert.Equal(t, 0, len(root.childQueueOrder))
+	assert.Equal(t, 1, root.NodeCount())
+	// 10 of 10 items have been dequeued
+	assert.Equal(t, 0, root.ItemCount())
+	assert.Equal(t, 1, root.NodeCount())
+
+	// assert nothing in local or child queues
+	assert.True(t, root.IsEmpty())
+}
+
+func TestEnqueueDuringDequeueRespectsRoundRobin(t *testing.T) {
+	root := NewTreeQueue("root", maxTestQueueLen)
+
+	cache := map[string]struct{}{}
+
+	// enqueue two items to path root:0
+	childPath := QueuePath{"0"}
+	item := makeQueueItemForChildPath(root, childPath, cache)
+	require.NoError(t, root.EnqueueBackByPath(childPath, item))
```

Comment:
Is there a reason why some tests in this file use the `require` package and others use the `assert` package? I'd prefer to use just one (usually `require`) unless there's a particular reason not to.

Reply:
I usually go with require for something that is unrecoverable, like a failure in test setup, in this case a failure to enqueue.
Then asserts for something where I might still gain information about what went wrong by continuing on after a failed assertion.

But I don't feel strongly so I have switched everything to require here.